project/profiles/capacity4more/modules/c4m/content/c4m_content/c4m_content.module

```diff
@@ -711,3 +721,190 @@ function c4m_content_create_highlight_link($gid, $nid, $highlight) {
 
   return l($star, $url, $options);
 }
+
+/**
+ * Build a tree of a hierarchical taxonomy and theme it.
+ *
+ * @param array $items
+ *   The items in the tree.
+ * @param array $link_term_to_facet
+ *   The current vocabulary machine name, for linking to facet.
+ *
+ * @return string
+ *   The markup of the tree.
+ */
+function c4m_content_taxonomy_build_tree($items, $link_term_to_facet = array()) {
+  // Each item at $trees is an array, representing two level tree.
+  // First node is tree root (level 1), and all others, if exist,
+  // are children (level 2).
+  $trees = array();
+
+  // Stores positions of trees to be shifted right, as they present
+  // second and third level, at 3 level taxonomy hierarchy.
+  $shift_right_array = array();
+
+  while (!empty($items)) {
+    // Extract first root item from $items array - level 1.
+    $root = _c4m_content_taxonomy_extract_root_item($items);
+    // Extract all roots children from $items array - level 2.
+    $children = _c4m_content_taxonomy_extract_child_items($items, $root->tid);
```

Comment:
@ordavidil 
Every iteration of while loop, extracts either two or three level tree from `$items` array.
If we get  `$root == FALSE` when `$items` is not empty, then we have dis-attached terms -> corrupted data, since terms are entered as trees.
Though such thing should not happen, I'll add a handler:
`if (!($root = _c4m_content_taxonomy_extract_root_item($items))) {
  break;
}`